### PR TITLE
TLS registry

### PR DESCRIPTION
**Pull request by projedi**

### CODE REVIEW

File: runtime/src/mm/cpp/ThreadLocalStorage.hpp [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+    using Key = void*;
+
+    class Iterator {
+    public:
+        explicit Iterator(std::vector<ObjHeader*>::iterator iterator) : iterator_(iterator) {}
+
+        ObjHeader** operator*() noexcept { return &*iterator_; }
+
+        Iterator& operator++() noexcept {
+          ++iterator_;
```
Comment (ilmat192) [inline runtime/src/mm/cpp/ThreadLocalStorage.hpp:30]: (minor) Indentation

**Review by ilmat192 — APPROVED** [review]

File: runtime/src/mm/cpp/ThreadLocalStorage.hpp (hunk 1 above)
Comment (projedi) [reply runtime/src/mm/cpp/ThreadLocalStorage.hpp]: Forgot to reformat, thanks!

File: runtime/src/mm/cpp/ThreadLocalStorage.hpp [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+    // Lookup value in storage. Can only be called after `Commit`.
+    ObjHeader** Lookup(Key key, int index) noexcept;
+
+    Iterator begin() noexcept { return Iterator(storage_.begin()); }
+    Iterator end() noexcept { return Iterator(storage_.end()); }
+
+private:
+    enum class State {
+        kBuilding,
+        kCommitted,
```
Comment (SvyatoslavScherbina) [inline runtime/src/mm/cpp/ThreadLocalStorage.hpp:56]: kCleared?

File: runtime/src/mm/cpp/ThreadLocalStorage.hpp [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+    };
+
+    ObjHeader** Lookup(Entry entry, int index) noexcept;
+
+    std::vector<ObjHeader*> storage_;
+    // TODO: `std::unordered_map` is probably the wrong container here.
+    std::unordered_map<Key, Entry> map_;
+    State state_ = State::kBuilding;
+    int size_ = 0; // Only used in `State::kBuilding`
+    std::pair<Key, Entry> lastEntry_;
```
Comment (SvyatoslavScherbina) [inline runtime/src/mm/cpp/ThreadLocalStorage.hpp:71]: lastKeyAndEntry_?

**Review by SvyatoslavScherbina — APPROVED** [review]

**Pull request merged**